### PR TITLE
Added the doze power mode support

### PR DESCRIPTION
**Pull request by pallavigopal**
Added the suport for the doze and doze suspend power modes.
For more details about doze and doze suspend refer the below link
https://developer.android.com/training/monitoring-device-state/doze-standby.html

Jira: IAHWC-21
Test: Not able to test the exact doze path device not entering into the idle mode
Signed-off-by: Pallavi G <pallavi.g@intel.com>

### CODE REVIEW

**Comment by pallavigopal** [comment]
Incorporated all the review comments

File: common/display/displayqueue.cpp [hunk 1]
```diff
@@ -149,27 +149,35 @@ bool DisplayQueue::ApplyPendingModeset(drmModeAtomicReqPtr property_set) {
...
+    dpms_mode_ = DRM_MODE_DPMS_OFF;
+    Flush();
+    drmModeConnectorSetProperty(gpu_fd_, connector_, dpms_prop_, dpms_mode_);
     Exit();
     return true;
   }
 
-  if (dpms_mode_ == DRM_MODE_DPMS_ON) {
+  if (power_mode == kDozeSuspend) {
+    Flush();
```
Comment (pallavigopal) [inline common/display/displayqueue.cpp]: do

File: public/hwcdefs.h [hunk 2]
```diff
@@ -63,5 +63,14 @@ enum class DisplayType : int32_t {
   kHeadless = 3
 };
 
+enum DisplayPowerMode {
+  kOff = 0,         // Display is off
```
Comment (pallavigopal) [inline public/hwcdefs.h]: i will do

File: common/display/displayqueue.cpp [hunk 3]
```diff
@@ -316,6 +344,8 @@ void DisplayQueue::HandleExit() {
     return;
   }
 
+  Flush();
```
Comment (pallavigopal) [inline common/display/displayqueue.cpp]: same like above

File: common/display/displayqueue.cpp [hunk 4]
```diff
@@ -296,18 +315,27 @@ void DisplayQueue::HandleRoutine() {
     ConditionalSuspend();
     return;
   }
+  display_queue_.unlock();
```
Comment (pallavigopal) [inline common/display/displayqueue.cpp]: yes to get the size of the queue_

File: common/display/displayqueue.cpp (hunk 1 above)
Comment (pallavigopal) [inline common/display/displayqueue.cpp]: same as above

File: common/display/displayqueue.cpp [hunk 5]
```diff
@@ -149,27 +149,35 @@ bool DisplayQueue::ApplyPendingModeset(drmModeAtomicReqPtr property_set) {
...
-  if (dpms_mode_ == dpms_mode)
+
+  if (dpms_mode_ == power_mode)
     return true;
 
-  dpms_mode_ = dpms_mode;
-  if (dpms_mode_ == DRM_MODE_DPMS_OFF) {
+  if ((power_mode == kOff) || (power_mode == kDoze)) {
+    dpms_mode_ = DRM_MODE_DPMS_OFF;
+    Flush();
```
Comment (pallavigopal) [inline common/display/displayqueue.cpp]: User experience wont be as expected if we skip those queued flips. I hope that is the expected user behavior we get the new flip if there is any geometry change in the incoming layers. please correct me if am wrong

Comment (kalyankondapally) [inline common/display/displayqueue.cpp]: The only worry I have is depending on the queue the power off/doze can be delayed a lot (If u think each update takes 16.6 msec and if we have two in queue than its 2* 16.6). If the action needs to be immediate than we need to do it immediately. But I think we need to test this out and see how it feels. Thats why a demo would be great to check out.

File: common/display/displayqueue.cpp [hunk 6]
```diff
@@ -215,8 +223,19 @@ bool DisplayQueue::QueueUpdate(std::vector<HwcLayer*>& source_layers) {
   return true;
 }
 
+void DisplayQueue::GetNextQueueItem(DisplayQueueItem& item) {
+  display_queue_.lock();
```
Comment (kalyankondapally) [inline common/display/displayqueue.cpp]: No need for the locks here. The callers are already holding the locks. Perhaps u want to separate this change to a separate commit and next commit can just use it. We can merge this cleanup in mean time. GetNextQueueItem being called from HandleUpdateRequest I mean.

Comment (pallavigopal) [inline common/display/displayqueue.cpp]: GetNextQueueItem is called from HandleRoutine not from HandleUpdateRequest.

File: common/display/displayqueue.cpp (hunk 5 above)
Comment (pallavigopal) [inline common/display/displayqueue.cpp]: Ideally the size of the queue should be zero. The off/doze/doze suspend will get called only when there is no activity in the current app/the device is in idle mode. Flush is just the ideal way of handling the queue exit just added to cover any corner case for some reason the flip delayed.
